README.md

```diff
@@ -0,0 +1,396 @@
+
+It is also a chance to try out some "interesting" ideas (such as config reducers) that would be much more difficult to sketch out in a real Agent.
+Maybe some of these are bad ideas, maybe they are things we can do later, or maybe we want to do them first thing.
+
+You can see the set of components in [`COMPONENTS.md`](./COMPONENTS.md).
+
+# Building
+
+To build, run `inv build`.
+This will build all binaries in the root of the repository.
+
+To regenerate files in the repo, run `inv generate`.
+
+# Component Guidelines
+
+What follows are draft guidelines for writing components.
+This repository follows these guidelines.
+
+If these guidelines are adopted, then the expectation is that all components would follow them.
+This reduces cognitive load when using components -- no need to remember one component's peculiarities.
+It also allows Agent-wide changes, where we make the same formulaic change to each component.
+If a situation arises that contradicts the guidelines, then we can update the guidelines (and change all affected components).
+
+## Prerequisites
+
+You should be familiar with [`fx`](https://pkg.go.dev/go.uber.org/fx), the tool used for dependency injection.
+
+## Component Definition
+
+A component is defined in a dedicated package under `comp/`, with the following defined in `component.go`:
+
+ * Extensive package-level documentation.
+   This should define, as precisely as possible, the behavior of the component, acting as a contract on which users of the component may depend.
+
+   The documentation (both package-level and method-level) should also include:
+
+   * Precise information about which interface methods can be called during the setup phase, and which must only be called after the component is started.
+   * Precise information about data ownership of passed values and returned values (does the method modify its arguments?  can the caller modify a returned slice or map?).
+   * Precise information about goroutines and blocking (does the method block? is a callback invoked in a dedicated goroutine? what happens if a channel is full?)
+
+ * `pkg.Component` -- the type implemented by the component.
+   This can be an empty interface, but is the type by which other components will find this one via `fx`.
+   It should have a formulaic doc string like `// Component is the component type.`, deferring documentation to the package docs.
+   All interface methods should be exported and thoroughly documented.
+
+ * `pkg.Module` -- an `fx.Option` that can be included in an `fx.App` to make this component available.
+   To improve logging, use `fx.Module(comppath, ..)`, where `comppath` is the root-relative package path, e.g., `"comp/util/log"`.
+   It should have a formulaic doc string like `// Module defines the fx options for this component.`
+
+Components should not be nested; that is, no component's Go path should be a prefix of another component's Go path.
```

Comment:
I have a few weak reasons:
 * It makes CODEOWNERS a little simpler (but this could be worked around with proper ordering)
 * It makes it a little clearer that components don't "contain" other components (but in a way that's what we want with bundles..)
 * No component has access to the `internal/`s of another component.

Reply:
Consider the following example of an `Aggregator` component that needs as dependency a `TimeSampler`.  Does `TimerSampler` is a component? 
- If yes, does it means there is a `comp/aggregator` and `comp/timesampler`?
- If no, does  `TimerSampler` is an internal component of `Aggregator` instead?

Reply:
I think both would be components, and they would probably be in the "dogstatsd" bundle.

Reply:
In this particular example: what would be the advantage of having `timesampler` has a component ? It's designed to be used by the aggregator exclusively.

I'm trying to better understand what will be the aim when pushing other teams to use component. Intuitively they're going to wrap their "main" folder in `pkg/` into a component and be done with it.

Reply:
The advantages of more granular components are:
 - abstraction for better understanding
 - isolation for testing
 - isolation of impact from changes to a component
 - regular treatment by other parts of the agent (health monitoring, for example)

In terms of DogStatsD, there's already a "component" structure in the diagrams.  But, digging into the code that structure is a bit murkier, and it's not clear what impact a change to, for example, the timesampler would have on the aggregator.  Bringing some engineering tools to bear will help clarify that structure: Go's exported items and interfaces to limit the interface between the components, explicit documentation of the behavior of each component, and tests for each component.

I suppose a team can decide to "cheat" and create one big component and move on.  The future velocity, bug load, retention, and onboarding effort in that team may be impacted :)

Reply:
Answering @ogaca-dd's comment regarding items 1-4.  That's a good breakdown of kinds of components!  Maybe 3 is where we let teams have some discretion.  Looking at the existing code, I see that a lot is already component-structured, but informally.  For example, AD has nicely-defined listener and config-poller components; WLM has listener components; dogstatsd has aggregator which in turn has a lot of [nice internal components](https://github.com/DataDog/datadog-agent/tree/main/pkg/aggregator#package-aggregator).  So, for some teams, it may be worth the investment to use the component architecture to represent those existing, informal components.  For others, working code may be good enough and the larger component can be left as-is.  Does that sound about right?